rfcs/text/0000-basic-typed-contracts.md

```diff
@@ -0,0 +1,249 @@
+applied to integers. Without the guard, `even?` might raise a runtime
+error (its own contract violation) unrelated to the contract the
+programmer wrote. To give types to these contracts, then, we need to
+know that whatever comes out of the `exact-integer?` contract is safe
+to flow into the `even?` contract. So in addition to needing to know
+the types of values a contract can be used on, we also need to know
+that what comes out of a contract is safe for some subsequent
+contract, like in `and/c`. Typed Racket's propositions help us uncover
+this last part.
+
+The two parts of a contract's type explain what the contract can be
+used on and what comes out of the contract. For `exact-integer?`, its
+function type says it can be used on `Any` and that it has a positive
+proposition for `Integer`, thus it has contract type `(Contract Any
+Integer)`. The type of `even?` is `(-> Integer Boolean)`, with no
+propositions, and so it has contract type `(Contract Integer
+Integer)`. We say that the first part of a contract's type is its
+*input type* and that the second part of a contract's type is its
+*output type*. For example, the input type of `exact-integer?` is
+`Any` and its output type is `Integer`.
+
+Back to type checking the combined contract, `(and/c exact-integer?
+even?)`. From the types above, we know that what comes out of
+`exact-integer?` can safely go in to `even?`. And since everything
+lines up, we take the the input type of the first contract type
+(`Integer`) and the output type of the last contract (`Integer`) and
+give the whole contract type `(Contract Any Integer)`.
+
+## Contracts on higher-order types
+
+If a function only consumes and produces positive reals, a Racket
+programmer might express that information using the contract `(->/c
+positive? positive?)`. In Typed Racket, that same information could be
+expressed with the type `(-> Positive-Real Positive-Real)`. So
+although that exact contract might not make sense in a Typed Racket
+program, we still want to give it a type---the path from Racket to
+Typed Racket isn't always smooth so we might be able to help the
+Racket programmer if we give a meaningful type to that contract.
+
+Starting with the function `positive?`, its type is `(-> Real
+Boolean)` and it has a positive proposition for `Positive-Real`. So
+based on what we saw in the previous section, it has the contract type
+`(Contract Real Positive-Real)`.
+
+The function contract combinator `->/c` wraps whatever function it is
+applied to, ensuring that a caller provides arguments that satisfy the
+domain contract and that the function returns a result that satisfies
+the range contract; Values flow from the caller through the domain
+contract into the function, and out of the function and through the
+range contract.
+
+This value flow is reflected in the type of a function contract. For
+our `(->/c positive? positive?)` the type is
+
+    (Contract (-> {2} Positive-Real Real {3})
```

Comment:
This notation is a little confusing. How about `(Contract (-> #|2|#Positive-Real #|3|#Real) ....)` so that:
1. it's valid TR syntax
2. the labels always come first
3. the labels touch the names they belong to

Reply:
Good idea. Will update the RFC